src/server/worker/stats.go

```diff
@@ -0,0 +1,115 @@
+package worker
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	// PrometheusPort is the port the aggregated metrics are served on for scraping
+	PrometheusPort = int32(9090)
+
+	datumCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "pachyderm",
+			Subsystem: "user",
+			Name:      "datum_count",
+			Help:      "Number of datums counted by pipeline and state",
+		},
+		[]string{
+			"pipeline",
+			"job",
+			"state",
+		},
+	)
+
+	bucketFactor  = 2.0
+	bucketCount   = 20 // Which makes the max bucket 2^20 seconds or ~12 days in size
+	datumProcTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "pachyderm",
+			Subsystem: "user",
+			Name:      "datum_proc_time",
+			Help:      "Time processing user code",
+			Buckets:   prometheus.ExponentialBuckets(1.0, bucketFactor, bucketCount),
+		},
+		[]string{
+			"pipeline",
+			"job",
+			"state", // Since both finished and errored datums can have proc times
+		},
+	)
+
+	datumDownloadTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "pachyderm",
+			Subsystem: "user",
+			Name:      "datum_download_time",
+			Help:      "Time to download input data",
+			Buckets:   prometheus.ExponentialBuckets(1.0, bucketFactor, bucketCount),
+		},
+		[]string{
+			"pipeline",
+			"job",
+		},
+	)
+
+	datumUploadTime = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "pachyderm",
+			Subsystem: "user",
+			Name:      "datum_upload_time",
+			Help:      "Time to upload output data",
+			Buckets:   prometheus.ExponentialBuckets(1.0, bucketFactor, bucketCount),
+		},
+		[]string{
+			"pipeline",
+			"job",
+		},
+	)
+
+	datumDownloadSize = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "pachyderm",
+			Subsystem: "user",
+			Name:      "datum_download_size",
+			Help:      "Size of downloaded input data",
+			Buckets:   prometheus.ExponentialBuckets(1.0, bucketFactor, bucketCount),
+		},
+		[]string{
+			"pipeline",
+			"job",
+		},
+	)
+
+	datumUploadSize = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "pachyderm",
+			Subsystem: "user",
+			Name:      "datum_upload_size",
+			Help:      "Size of uploaded output data",
+			Buckets:   prometheus.ExponentialBuckets(1.0, bucketFactor, bucketCount),
+		},
+		[]string{
+			"pipeline",
+			"job",
+		},
+	)
```

Comment:
Why are all of these Histograms? They all seem like things that could in theory be counters? I'm guessing there's some queries you couldn't do with counters, but it might make sense to include the counter version as well since it's more basic and I think there are some useful things we could do with it.